src/main/scala/is/hail/rvd/RVD.scala

```diff
@@ -1,27 +1,143 @@
 package is.hail.rvd
 
-import is.hail.annotations.{Region, RegionValue, RegionValueBuilder}
-import is.hail.expr.types.Type
+import is.hail.HailContext
+import is.hail.annotations._
+import is.hail.expr.{JSONAnnotationImpex, Parser}
+import is.hail.expr.types.{TArray, TStruct}
+import is.hail.io.RichRDDRegionValue
 import is.hail.utils._
 import org.apache.spark.{Partition, SparkContext}
 import org.apache.spark.rdd.{AggregateWithContext, RDD}
+import org.apache.spark.sql.Row
 import org.apache.spark.storage.StorageLevel
+import org.json4s.JValue
+import org.json4s.JsonAST.{JArray, JObject, JString}
 
 import scala.reflect.ClassTag
 
-object RVD {
-  def apply(rowType: Type, rdd: RDD[RegionValue]): ConcreteRVD = new ConcreteRVD(rowType, rdd)
+object RVDSpec {
+  def extract(jv: JValue): RVDSpec = {
+    jv \ "name" match {
+      case JString("UnpartitionedRVDSpec") =>
+        UnpartitionedRVDSpec.extract(jv \ "args")
+      case JString("OrderedRVDSpec") =>
+        OrderedRVDSpec.extract(jv \ "args")
+    }
+  }
 
-  def empty(sc: SparkContext, rowType: Type) = ConcreteRVD.empty(sc, rowType)
+  def readLocal(hc: HailContext, path: String, rowType: TStruct, partFiles: Array[String]): IndexedSeq[Row] = {
+    assert(partFiles.length == 1)
+
+    val hConf = hc.hadoopConf
+    partFiles.flatMap { p =>
+      val f = path + "/parts/" + p
+      val in = hConf.unsafeReader(f)
+      HailContext.readRowsPartition(rowType)(0, in)
+        .map(rv => new UnsafeRow(rowType, rv.region.copy(), rv.offset))
+    }
+      .toFastIndexedSeq
+  }
+}
+
+abstract class RVDSpec {
+  def read(hc: HailContext, path: String): RVD
+
+  def readLocal(hc: HailContext, path: String): IndexedSeq[Row]
+
+  def toJSON: JValue
+}
+
+object UnpartitionedRVDSpec {
+
+  case class JSONArgs(row_type: String, part_files: Array[String])
+
+  def extract(jargs: JValue): UnpartitionedRVDSpec = {
+    val args = jargs.extract[JSONArgs]
+    UnpartitionedRVDSpec(Parser.parseStructType(args.row_type), args.part_files)
+  }
+}
```

Comment:
You can use a custom serializer in json4s to avoid repeating the matrix parser
and TStruct parser stuff.

```scala
class TStructSerializer extends CustomSerializer[TStruct](format => (
  { case JString(s) => Parser.parseStructType(args.row_type) },
  { case x: TStruct => x.toString() }
))
```
you have to have the right formats in your implicit scope:
```scala
implicit val formats = DefaultFormats + new TStructSerializer
```

but then this is just:
```scala
object UnpartitionedRVDSpec {
  case class JSONArgs(row_type: String, part_files: Array[String])

  def extract(jargs: JValue): UnpartitionedRVDSpec = {
    val args = jargs.extract[JSONArgs]
    UnpartitionedRVDSpec(args.row_type, args.part_files)
  }
}

case class UnpartitionedRVDSpec(
  rowType: TStruct,
  partFiles: Array[String]) extends RVDSpec {
  def read(hc: HailContext, path: String): UnpartitionedRVD =
    new UnpartitionedRVD(rowType, hc.readRows(path, rowType, partFiles))

  def readLocal(hc: HailContext, path: String): IndexedSeq[Row] =
    RVDSpec.readLocal(hc, path, rowType, partFiles)

  def toJSON: JValue = JObject("name" -> JString("UnpartitionedRVDSpec"),
    "args" -> render(UnpartitionedRVDSpec.JSONArgs(rowType, partFiles)))
}
```

*but* we can get even better using TypeHints:

```scala
class TStructSerializer extends CustomSerializer[TStruct](format => (
  { case JString(s) => Parser.parseStructType(args.row_type) },
  { case x: TStruct => x.toString() }
))

implicit val formats = new DefaultFormats() {
  override val typeHints = ShortTypeHints(List(classOf[Foo]))
  override val typeHintFieldName = "name" } + new TStructSerializer

object UnpartitionedRVDSpec {
  def extract(jargs: JValue): UnpartitionedRVDSpec =
    jargs.extract[UnpartitionedRVDSpec]
}

case class UnpartitionedRVDSpec(
  row_type: TStruct,
  part_files: Array[String]) extends RVDSpec {
  def read(hc: HailContext, path: String): UnpartitionedRVD =
    new UnpartitionedRVD(rowType, hc.readRows(path, row_type, part_files))

  def readLocal(hc: HailContext, path: String): IndexedSeq[Row] =
    RVDSpec.readLocal(hc, path, row_type, part_files)

  def toJSON: JValue = render(this)
}
```

And the JSON should look like:

```javascript
{ "name" : "UnpartitionedRVDSpec"
, "row_type" : "... the toString'ed row type ..."
, "part_files" : ["part-00000", ...]
}
```

This seems to save us some of the boilerplate.